content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md

```diff
@@ -0,0 +1,249 @@
+```
+
+If you don't, you could see such messages along the way:
+
+```
+===> DETECTING
+======== Error: paketo-buildpacks/spring-boot@{{.version}} ========
+fork/exec /cnb/buildpacks/paketo-buildpacks_spring-boot/{{.version}}/bin/detect: exec format error
+```
+
+### Run the Paketo Buildpack
+
+If you try to build a [sample Java project](https://github.com/paketo-buildpacks/samples/tree/main/java/maven) using your newly build `spring-boot` buildpack:
+
+```shell
+pack build -p ~/workspaces/paketo-buildpacks/samples/java/maven applications/maven-app \
+  -b .
+```
+`-p` providing the path of the application to build an image for.
+
+`-b .` providing the path of our buildpack we just built (it will be looking for `bin/detect` and `bin/build` under this folder)
+
+Well, you will get an error:
+
+```
+===> DETECTING
+======== Results ========
+pass: paketo-buildpacks/spring-boot@{{.version}}
+Resolving plan... (try #1)
+fail: paketo-buildpacks/spring-boot@{{.version}} requires jvm-application
+```
+
+With Paketo Buildpacks, you typically can't run just one buildpack; most of them are inter-dependent.
+
+In our case, `spring-boot` buildpack `requires` a buildpack providing `jvm-application`; probably the `executable-jar` buildpack will provide it, but then this one will `require` a buildpack providing a JVM, etc.
+
+#### Note
+
+You would think that `-b .` would just append our buildpack after the other detected buildpacks provided by the builder... Unfortunately, as soon as you use `-b` once, `pack` will only consider the buildpacks provided in the command line.
+
+There's a way not to spend hours to find the right incantation: observing what the chain is when you let your builder pick the needed `buildpacks`:
+
+```
+pack build -p ~/workspaces/paketo-buildpacks/samples/java/maven applications/maven-app
+[...]
+===> DETECTING
+10 of 26 buildpacks participating
+paketo-buildpacks/ca-certificates   3.6.1
+paketo-buildpacks/bellsoft-liberica 10.2.3
+paketo-buildpacks/syft              1.30.0
+paketo-buildpacks/maven             6.15.2
+paketo-buildpacks/executable-jar    6.7.1
+paketo-buildpacks/apache-tomcat     7.13.3
+paketo-buildpacks/apache-tomee      1.7.1
+paketo-buildpacks/liberty           3.7.1
+paketo-buildpacks/dist-zip          5.6.1
+paketo-buildpacks/spring-boot       5.25.0
+[...]
+```
+
+Now, that we know the full chain, we just need to replace the original `paketo-buildpacks/spring-boot` with ours.
+
+```
+pack build -p ~/workspaces/paketo-buildpacks/samples/java/maven applications/maven-app \
+  -b paketo-buildpacks/ca-certificates \
+  -b paketo-buildpacks/bellsoft-liberica \
+  -b paketo-buildpacks/syft \
+  -b paketo-buildpacks/maven \
+  -b paketo-buildpacks/executable-jar \
+  -b paketo-buildpacks/apache-tomcat \
+  -b paketo-buildpacks/apache-tomee \
+  -b paketo-buildpacks/liberty \
+  -b paketo-buildpacks/dist-zip \
+  -b . 
```

Comment:
TIP: If you prepend `urn:cnb:builder:` to the buildpacks you are not modifying then `pack` will take them from the builder. If you do not, then pack will dynamically add all of those buildpacks to the builder. It is slightly faster to use the ones already in the builder.